app/utility/base_knowledge_svc.py

```diff
@@ -0,0 +1,345 @@
+        :param criteria: dictionary containing fields to match on
+        """
+        return self._remove('relationships', criteria)
+
+    # --- Rule API ---
+
+    def _add_rule(self, rule, constraints=None):
+        """
+        Add a rule to the internal store
+        :param rule: Rule object to add
+        :param constraints: dictionary containing fields to match on
+        """
+        ###
+        #    rule.action: [DENY, ALLOW, EXCLUSIVE, EXCLUSIVE_TRAIT, EXCLUSIVE_VALUE], 'EXCLUSIVE_*' actions denote that
+        #        the trait/value will be made mutually exclusive in its use to the agent/group/operation that is
+        #        specified for. Essentially a fact is binded to mutex, and only one action can be using the fact
+        #        at any one time.
+        ###
+        if not any((x.action == rule.action) and (x.trait == rule.trait) for x in self.fact_ram['rules']):
+            rule._knowledge_id = uuid.uuid4()
+            self.fact_ram['rules'].append(rule)
+            if constraints:
+                self.fact_ram['constraints'][rule._knowledge_id] = constraints
+
+    def _get_rules(self, criteria, restrictions=None):
+        """
+        Retrieve rules from the internal store
+        :param criteria: dictionary containing fields to match on
+        :return: list of matching rules
+        """
+        complete_list = [x for x in self.fact_ram['rules'] if self._check_rule(x, criteria, True)]
+        return self._apply_restrictions(complete_list, restrictions)
+
+    def _delete_rule(self, criteria):
+        """
+        Remove a rule from the internal store
+        :param criteria: dictionary containing fields to match on
+        """
+        sublist = [x for x in self.fact_ram['rules'] if self._check_rule(x, criteria)]
+        self._clear_matching_constraints(sublist)
+        self.fact_ram['rules'][:] = [x for x in self.fact_ram['rules'] if x not in sublist]
+        return
+
+    @staticmethod
+    def _check_rule(rule, desired_quals, wildcard=False):
+        """
+        Check whether or not a rule matches the provided criteria
+        :param rule: Rule object to check
+        :param desired_quals: dictionary containing fields to match on
+        :param wildcard: whether or not to do wildcard matching on the 'match' field
+        :return:
+        """
+        criteria_matches = []
+        for k, v in desired_quals.items():
+            if getattr(rule, k) == v:
+                criteria_matches.append(True)
+            elif k == 'match' and wildcard:  # support wildcard matching
+                val = getattr(rule, k)
+                local_stub = val.split('*')
+                input_stub = v.split('*')
+                if all(x in v for x in local_stub) or all(x in val for x in input_stub):
+                    criteria_matches.append(True)
+        if len(criteria_matches) == len(desired_quals):
+            return True
+
+    # --- New Inferencing API ---
+    # NOT IMPLEMENTED YET
```

Comment:
Yeah, it kind of is. I left this in because the entire API is not yet implemented, while the NotImplementedErrors indicate the individual methods aren't ready to go yet. I can remove the line if you'd prefer.